datumaro/components/project.py

```diff
@@ -2,486 +2,2389 @@
+        self.__dict__['_config'] = config
+        self.__dict__['_readonly'] = readonly
+        self.__dict__['name'] = source
 
-            def get(self, id, subset=None):
-                subset = subset or self.name
-                assert subset == self.name, '%s != %s' % (subset, self.name)
-                return super().get(id, subset)
+    def save(self, save_dir=None, **kwargs):
+        if save_dir is None and self.readonly:
+            raise ReadonlyDatasetError()
+        self._dataset.save(save_dir, **kwargs)
 
-    def __init__(self, project):
-        super().__init__()
+    @property
+    def readonly(self):
+        return self._readonly or not self.is_bound
+
+    @property
+    def config(self):
+        return self._config
+
+    def __getattr__(self, name):
+        return getattr(self._dataset, name)
+
+    def __setattr__(self, name, value):
+        return setattr(self._dataset, name, value)
 
+    def __iter__(self):
+        yield from self._dataset
+
+    def __len__(self):
+        return len(self._dataset)
+
+    def subsets(self):
+        return self._dataset.subsets()
+
+    def get_subset(self, name):
+        return self._dataset.get_subset(name)
+
+    def categories(self):
+        return self._dataset.categories()
+
+    def get(self, id, subset=None):
+        return self._dataset.get(id, subset)
+
+
+class IgnoreMode(Enum):
+    rewrite = auto()
+    append = auto()
+    remove = auto()
+
+def _update_ignore_file(paths: Union[str, List[str]], repo_root: str,
+        filepath: str, mode: Optional[IgnoreMode] = None):
+    def _make_ignored_path(path):
+        path = osp.join(repo_root, osp.normpath(path))
+        assert path.startswith(repo_root), path
+        return osp.relpath(path, repo_root)
+
+    header = '# The file is autogenerated by Datumaro'
+
+    mode = parse_str_enum_value(mode, IgnoreMode, IgnoreMode.append)
+
+    if isinstance(paths, str):
+        paths = [paths]
+    paths = {osp.join(repo_root, osp.normpath(p)): _make_ignored_path(p)
+        for p in paths}
+
+    openmode = 'r+'
+    if not osp.isfile(filepath):
+        openmode = 'w+' # r+ cannot create, w+ truncates
+    with open(filepath, openmode) as f:
+        lines = []
+        if mode in {IgnoreMode.append, IgnoreMode.remove}:
+            for line in f:
+                lines.append(line.strip())
+            f.seek(0)
+
+        new_lines = []
+        for line in lines:
+            if not line or line.startswith('#'):
+                new_lines.append(line)
+                continue
+
+            line_path = osp.join(repo_root,
+                osp.normpath(line.split('#', maxsplit=1)[0]).lstrip('/'))
+
+            if mode == IgnoreMode.append:
+                if line_path in paths:
+                    paths.pop(line_path)
+                new_lines.append(line)
+            elif mode == IgnoreMode.remove:
+                if line_path in paths:
+                    paths.pop(line_path)
```

Comment:
I think this will work incorrectly if the same path is listed twice in the ignore file (I assume that the ignore file is conceptually a set, and thus a path listed twice has the same meaning as it being listed once).

Reply:
Probably, it is right, but there is no practical reason for this, so if it happens, the user will probably know who to blame.

Reply:
But why not fix it? All you would need to do is to remove the `pop` call.

Reply:
Done